llvm/lib/IR/RepoTicket.cpp

```diff
@@ -118,76 +127,122 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
+                           NullDigest);
+
+  const auto NumVisited = Visited.size();
+  LLVM_DEBUG(dbgs() << "Computing hash for \"" << GO->getName() << "\" (#"
+                    << NumVisited << ")\n");
+
+  // If the hash has been memorized, we can return the result immediately.
+  const auto TablePos = HashTable.find(GO);
+  if (TablePos != HashTable.end()) {
+    ++VisitedCachedGOTimes;
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(true, NumVisited, TablePos->second);
+  }
+
+  MD5 GOHash = MD5();
 
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, NumVisited);
   if (!Inserted) {
+    assert(NumVisited > StateIt->second);
     // If GO is visited, use the letter 'R' as the marker and use its state as
     // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    GOHash.update(static_cast<uint8_t>(Tags::Backref));
+    GOHash.update(NumVisited - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    MD5::MD5Result Digest;
+    GOHash.final(Digest);
+    return std::make_tuple(true, StateIt->second, Digest);
   }
 
   bool Changed = false;
-  GOHash.update('T');
+  GOHash.update(static_cast<uint8_t>(Tags::GO));
   auto State = AccumulateGODigest(GO, GOHash, GOIMap);
   Changed = Changed || State.Changed;
-  // Update the GO's hash using its contributions and dependencies.
-  auto &GOContributions = State.Contributions;
-  auto &GODependencies = State.Dependencies;
-  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
-           make_range(GOContributions.begin(), GOContributions.end()),
-           make_range(GODependencies.begin(), GODependencies.end()))) {
+  // Update the GO's hash using its contributions. Contributions don't form the
+  // loop since 1) only the global variable has the contributions and 2) the
+  // contributions are the functions.
+  for (const GlobalObject *const G : State.Contributions) {
+    Changed = true;
+    GOHash.update(GOIMap[G].InitialDigest.Bytes);
+  }
+  // Update the GO's hash using its dependencies.
+  auto LoopPoint = std::numeric_limits<size_t>::max();
+  for (const GlobalObject *const G : State.Dependencies) {
     const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined and not be discarded if it is not used,
     // skip it.
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
         !Fn->isDiscardableIfUnused())
       continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
+    const auto AdjGO = updateDigestUseDependenciesAndContributions(
```

Comment:
What is an AdjGo?

Reply:
Renamed to `GState`, which represents the state after accumulated G’s digest.